pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,604 @@
+	cloned.downstreamVector = cnVector.Window(vecDup, offset, offset+length, cloned.downstreamVector)
+
+	return cloned
+}
+
+func (vec *CnTaeVector[T]) Window(offset, length int) Vector {
+
+	// In DN Vector, we are using SharedReference for Window.
+	// In CN Vector, we are creating a new Clone for Window.
+	// So inorder to retain the nature of DN vector, we had use CnTaeVectorWindow Adapter.
+	return &CnTaeVectorWindow[T]{
+		ref: vec,
+		windowBase: &windowBase{
+			offset: offset,
+			length: length,
+		},
+	}
+}
+
+// TODO: --- We can remove below functions as they don't have any usage
+
+func (vec *CnTaeVector[T]) IsView() bool {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) GetView() VectorView {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) DataWindow(offset, length int) []byte {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) Data() []byte {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) SlicePtr() unsafe.Pointer {
+	slice := vec.Slice().([]T)
+	return unsafe.Pointer(&slice[0])
+}
+
+func (vec *CnTaeVector[T]) AppendNoNulls(s any) {
+	panic("Soon Deprecated")
+}
+
+// TODO: --- We can remove below function as they are only used in Testcases.
+
+func (vec *CnTaeVector[T]) Delete(delRowId int) {
+	deletes := roaring.BitmapOf(uint32(delRowId))
+	vec.Compact(deletes)
+}
+
+func (vec *CnTaeVector[T]) ReadFromFile(f common.IVFile, buffer *bytes.Buffer) (err error) {
+	stat := f.Stat()
+	var n []byte
+	var buf []byte
+	var tmpNode []byte
+	if stat.CompressAlgo() != compress.None {
+		osize := int(stat.OriginSize())
+		size := stat.Size()
+		tmpNode, err = vec.GetAllocator().Alloc(int(size))
+		if err != nil {
+			return
+		}
+		defer vec.GetAllocator().Free(tmpNode)
+		srcBuf := tmpNode[:size]
+		if _, err = f.Read(srcBuf); err != nil {
+			return
+		}
+		if buffer == nil {
+			n, err = vec.GetAllocator().Alloc(osize)
+			if err != nil {
+				return
+			}
+			buf = n[:osize]
+		} else {
+			buffer.Reset()
+			if osize > buffer.Cap() {
+				buffer.Grow(osize)
+			}
+			buf = buffer.Bytes()[:osize]
+		}
+		if _, err = compress.Decompress(srcBuf, buf, compress.Lz4); err != nil {
+			if n != nil {
+				vec.GetAllocator().Free(n)
+			}
+			return nil
+		}
+	}
+
+	_, err = vec.ReadFrom(bytes.NewBuffer(buf))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TODO: --- Below Functions Can be implemented in CN Vector.
+
+func (vec *CnTaeVector[T]) Equals(o Vector) bool {
+
+	if vec.Length() != o.Length() {
+		return false
+	}
+	if vec.GetType() != o.GetType() {
+		return false
+	}
+	if vec.Nullable() != o.Nullable() {
+		return false
+	}
+	if vec.HasNull() != o.HasNull() {
+		return false
+	}
+	if vec.HasNull() {
+		if !vec.NullMask().Equals(o.NullMask()) {
+			return false
+		}
+	}
+	mask := vec.NullMask()
+	for i := 0; i < vec.Length(); i++ {
+		if mask != nil && mask.ContainsInt(i) {
+			continue
+		}
+		var v T
+		if _, ok := any(v).([]byte); ok {
```

Comment:
these type assertion within a loop is very fishy -- at least we should be able to type cast only once outside of the loop.

Reply:
Sure. I will look into this.

Reply:
@aunjgr Can Equals be implemented in CN Vector?